factory/src/main/kotlin/io/qalipsis/core/factory/meters/inmemory/InMemoryCumulativeMeterRegistry.kt

```diff
@@ -0,0 +1,191 @@
+/*
+ * QALIPSIS
+ * Copyright (C) 2024 AERIS IT Solutions GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package io.qalipsis.core.factory.meters.inmemory
+
+import io.aerisconsulting.catadioptre.KTestable
+import io.qalipsis.api.logging.LoggerHelper.logger
+import io.qalipsis.api.meters.Counter
+import io.qalipsis.api.meters.DistributionSummary
+import io.qalipsis.api.meters.Gauge
+import io.qalipsis.api.meters.Meter
+import io.qalipsis.api.meters.MeterSnapshot
+import io.qalipsis.api.meters.Timer
+import io.qalipsis.core.factory.campaign.Campaign
+import io.qalipsis.core.factory.campaign.CampaignLifeCycleAware
+import io.qalipsis.core.factory.meters.CompositeCounter
+import io.qalipsis.core.factory.meters.CompositeDistributionSummary
+import io.qalipsis.core.factory.meters.CompositeGauge
+import io.qalipsis.core.factory.meters.CompositeTimer
+import io.qalipsis.core.factory.meters.MeterRegistry
+import io.qalipsis.core.reporter.MeterReporter
+import jakarta.inject.Singleton
+import java.time.Instant
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Default implementation of [MeterRegistry] that stores the [Meter]s into memory.
+ */
+@Singleton
+internal class InMemoryCumulativeMeterRegistry(
+    private val meterReporter: MeterReporter,
+) : MeterRegistry, CampaignLifeCycleAware {
+
+    /**
+     * Contains all the unique meters.
+     */
+    @KTestable
+    private val meters = ConcurrentHashMap<Meter.Id, Meter<*>>()
+
+    private var snapshotsConverter: (Collection<MeterSnapshot>) -> Collection<MeterSnapshot> = NO_OP_SNAPSHOTS_CONVERTER
+
+    private var meterConverter: MeterConverter = NO_OP_SCENARIO_METER_CONVERTER
+
+    override suspend fun init(campaign: Campaign) {
+        meters.clear()
+        super.init(campaign)
+        if (campaign.scenarios.size == 1) {
+            // When there is a unique scenario running in the campaign, we let the meters being created normally,
+            // but when the snapshots are generated, the ones attached to a
```

Comment:
I am not sure I get the point here

Reply:
Thank you, the documentation was not complete. What do you think now?

Reply:
Perfect now